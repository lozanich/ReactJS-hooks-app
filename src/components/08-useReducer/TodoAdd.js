import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({ handleAddToDo }) => {
  const { values, handleInputChange, reset } = useForm({
    description: "",
  });

  const { description } = values;

  const handleSubmit = (e) => {
    e.preventDefault();
    if (description.trim().length <= 1) {
      return;
    }

    const newTask = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };

    handleAddToDo(newTask);
    reset();
  };

  return (
    <>
      <h4>Agregar TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          onChange={handleInputChange}
          value={description}
          name="description"
          placeholder="Aprender../"
          autoComplete="off"
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Agregar
        </button>
      </form>
    </>
  );
};
