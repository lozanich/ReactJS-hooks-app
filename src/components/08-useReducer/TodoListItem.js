import React from "react";
import PropTypes from "prop-types";

export const TodoListItem = ({ item, index, handleDelete, handleToggle }) => {
  return (
    <li key={item.id} className="list-group-item">
      <p
        className={`${item.done && "complete"}`}
        onClick={() => handleToggle(item)}
      >
        {index + 1}. {item.desc}
      </p>

      <button onClick={() => handleDelete(item)} className="btn btn-danger">
        Eliminar
      </button>
    </li>
  );
};

TodoListItem.propTypes = {
  item: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleToggle: PropTypes.func.isRequired,
};
