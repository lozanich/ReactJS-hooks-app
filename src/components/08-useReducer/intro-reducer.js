console.log("Hello world");

const initialState = [
  {
    id: 1,
    todo: "Comprar leche",
    done: false,
  },
];

const todoReducer = (state = initialState, action) => {
  if (action?.type === "add") {
    return [...state, action.payload];
  }
  return state;
};

let todos = todoReducer();

const newTodo = {
  id: 2,
  todo: "Comprar leche",
  done: false,
};

const actionAdd = {
  type: "add",
  payload: newTodo,
};

todos = todoReducer(todos, actionAdd);

console.log(todos);
