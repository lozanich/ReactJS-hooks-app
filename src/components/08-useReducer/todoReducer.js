export const todoReducer = (state = [], action) => {
  switch (action.type) {
    case "add":
      return [...state, action.payload];

    case "delete":
      return state.filter((i) => i.id !== action.payload.id);

    case "toggle":
      // console.log("toggle");
      // console.log(action.payload);
      return state.map((item) =>
        item.id === action.payload.id ? { ...item, done: !item.done } : item
      );

    default:
      return state;
  }
};
