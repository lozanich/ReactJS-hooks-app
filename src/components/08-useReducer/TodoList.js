import React from "react";
import { TodoListItem } from "./TodoListItem";
import PropTypes from "prop-types";

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
  return (
    <ul className="list-group list-group-flush">
      {todos.map((item, id) => {
        return (
          <TodoListItem
            key={item.id}
            item={item}
            index={id}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        );
      })}
    </ul>
  );
};

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handleToggle: PropTypes.func.isRequired,
};
