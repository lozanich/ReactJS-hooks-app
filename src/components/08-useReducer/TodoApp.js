import React, { useEffect, useReducer } from "react";
import { todoReducer } from "./todoReducer";
import { TodoList } from "./TodoList";
import { TodoAdd } from "./TodoAdd";
import "./todoApp.css";

const init = () => {
  return JSON.parse(localStorage.getItem("tasks")) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  useEffect(() => {
    localStorage.setItem("tasks", JSON.stringify(todos));
  }, [todos]);
  // console.log(todos);

  const handleDelete = (item) => {
    const actionDelete = {
      type: "delete",
      payload: item,
    };
    dispatch(actionDelete);
  };

  const handleTaskCompleted = (item) => {
    const actionTaskCompleted = {
      type: "toggle",
      payload: item,
    };
    dispatch(actionTaskCompleted);
  };

  const handleAddTodo = (newToDo) => {
    dispatch({
      type: "add",
      payload: newToDo,
    });
  };

  return (
    <>
      <h1>Todo App ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-sm-6">
          {todos.length > 0 && (
            <TodoList
              todos={todos}
              handleDelete={handleDelete}
              handleToggle={handleTaskCompleted}
            />
          )}
        </div>

        <div className="col-sm-6">
          <TodoAdd handleAddToDo={handleAddTodo} />
        </div>
      </div>
    </>
  );
};
