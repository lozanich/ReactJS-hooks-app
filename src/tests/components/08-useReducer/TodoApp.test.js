import React from "react";
import "@testing-library/jest-dom";
import { mount, shallow } from "enzyme";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { demoTodos } from "../../fixtures/demoTodos";
import { act } from "@testing-library/react";

describe("Pruebas en el componente <TodoApp />", () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn(() => {});

  test("Debe de coincidir con el snapshot ", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de agregaar un TODO", () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find("TodoAdd").prop("handleAddToDo")(demoTodos[0]);
      wrapper.find("TodoAdd").prop("handleAddToDo")(demoTodos[1]);
    });

    expect(wrapper.find("h1").text().trim()).toBe("Todo App (2)");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test("Debe de Eliminar un TODO", () => {
    wrapper.find("TodoAdd").prop("handleAddToDo")(demoTodos[0]);
    wrapper.find("TodoList").prop("handleDelete")(demoTodos[0]);

    expect(wrapper.find("h1").text().trim()).toBe("Todo App (0)");
  });
});
