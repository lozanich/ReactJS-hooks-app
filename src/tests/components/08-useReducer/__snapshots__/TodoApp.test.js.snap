// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoApp /> Debe de coincidir con el snapshot  1`] = `
<Fragment>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-6"
    />
    <div
      className="col-sm-6"
    >
      <TodoAdd
        handleAddToDo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
